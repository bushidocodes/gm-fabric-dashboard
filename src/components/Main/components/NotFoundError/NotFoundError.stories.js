import React from "react";
import { storiesOf } from "@storybook/react";
import { withKnobs, text } from "@storybook/addon-knobs/react";

import NotFoundError from "./NotFoundError";

storiesOf("NotFoundError", module)
  .addDecorator(withKnobs)
  .add("default", () => (
    <div style={{ display: "flex" }}>
      <NotFoundError errorMsg={text("errorMsg")} />
    </div>
  ));
