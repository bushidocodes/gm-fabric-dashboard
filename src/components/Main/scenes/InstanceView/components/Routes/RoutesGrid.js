import { PropTypes } from "prop-types";
import React, { Component, Fragment } from "react";
import { connect } from "react-redux";
import _ from "lodash";

import Table from "components/Main/components/Table";
import TableToolbar from "components/Main/components/TableToolbar";
import ErrorBoundary from "components/ErrorBoundary";
import NotFoundError from "components/Main/components/NotFoundError";

import { getRoutesTable } from "utils/go/selectors";
import withUrlState from "components/withUrlState";

/**
 * Routes Container
 * Parent of RoutesTable and RoutesTableToolbar
 * Contains sort and filter logic for RoutesTable
 * @class RoutesGrid
 * @extends {Component}
 */

class RoutesGrid extends Component {
  static propTypes = {
    routes: PropTypes.array,
    setUrlState: PropTypes.func.isRequired,
    urlState: PropTypes.object.isRequired
  };

  /**
   * Helper function that takes the routes passed as props and sorts according to how sortByAttribute and ascending is set
   * in the url bar.
   * @param {Array} routes
   */
  sort = routes => {
    const {
      ascending = "true",
      sortByAttribute = "route"
    } = this.props.urlState;
    const sortOrder = ascending === "true" ? ["asc"] : ["desc"];

    return _.orderBy(routes, sortByAttribute, sortOrder);
  };

  /**
   * Helper function used internally to either sort by the key if not yet used to sort or toggle ascending / descending
   * if the key is already active.
   * When sorting by latency and error percent and the sortKey is new, set the initial sort order to descending
   * @param {String} newSortByAttribute
   */
  setSortByAttribute = newSortByAttribute => {
    const {
      urlState: { ascending = "true", sortByAttribute = "route" },
      setUrlState
    } = this.props;

    if (sortByAttribute === newSortByAttribute) {
      setUrlState({
        ascending: !JSON.parse(ascending)
      });
    } else if (
      newSortByAttribute === "errorPercent" ||
      newSortByAttribute.includes("latency")
    ) {
      setUrlState({
        sortByAttribute: newSortByAttribute,
        ascending: false
      });
    } else {
      setUrlState({
        sortByAttribute: newSortByAttribute
      });
    }
  };

  render() {
    const {
      setUrlState,
      urlState: { filterString = "", sortByAttribute = "route" },
      routes
    } = this.props;

    if (routes && routes.length > 0) {
      return (
        <Fragment>
          <TableToolbar
            searchInputProps={{
              filterString,
              setFilterString: filterString => setUrlState({ filterString }),
              searchPlaceholder: "Search Routes"
            }}
            sortByProps={{
              sortByOptions: [
                {
                  value: "route",
                  label: "Route"
                },
                {
                  value: "requests",
                  label: "Requests"
                },
                {
                  value: "errorPercent",
                  label: "Error %"
                },
                {
                  value: "latency50",
                  label: "Latency 50%"
                },
                {
                  value: "latency99",
                  label: "Latency 99%"
                }
              ],
              sortByAttribute,
              setSortByAttribute: this.setSortByAttribute
            }}
          />
          <ErrorBoundary>
            <Table
              type={"Route"}
              items={this.sort(
                this.props.routes.filter(
                  routeObj =>
                    routeObj.route
                      .toLowerCase()
                      .indexOf(filterString.trim().toLowerCase()) !== -1
                )
              )}
            />
          </ErrorBoundary>
        </Fragment>
      );
    } else {
      return <NotFoundError errorMsg={"No Routes Found"} />;
    }
  }
}

function mapStateToProps(state) {
  return {
    routes: getRoutesTable(state)
  };
}

export default connect(mapStateToProps)(withUrlState()(RoutesGrid));
