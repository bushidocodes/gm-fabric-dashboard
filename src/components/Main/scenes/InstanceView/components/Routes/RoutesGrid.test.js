import React from "react";
import { mount } from "enzyme";
import configureMockStore from "redux-mock-store";
import renderer from "react-test-renderer";
import NotFoundError from "components/Main/components/NotFoundError";

// Utilities
import mockState from "json/mockReduxState";

// Components
import RoutesGrid from "./index";

// Create a mock store and initialize with mock data
const store = configureMockStore()(mockState);

// Props that should be passed down to routes Table
const routesTableProps = {
  items: [
    {
      errorPercent: "0.000",
      errorsCount: 0,
      inThroughput: 0,
      latency50: 0,
      latency99: 9,
      outThroughput: 1580,
      requests: 17725,
      requestsPerSecond_dygraph: [
        [
          [new Date("2017-07-18T22:13:34.314Z"), 0],
          [new Date("2017-07-18T22:13:49.215Z"), 0],
          [new Date("2017-07-18T22:14:04.215Z"), 0],
          [new Date("2017-07-18T22:14:19.217Z"), 0],
          [new Date("2017-07-18T22:14:34.215Z"), 0],
          [new Date("2017-07-18T22:14:49.215Z"), 0],
          [new Date("2017-07-18T22:15:04.216Z"), 0],
          [new Date("2017-07-18T22:15:19.215Z"), 0],
          [new Date("2017-07-18T22:15:34.216Z"), 0]
        ],
        { labels: ["time", "Requests"] }
      ],
      requestsPerSecond_sparkline: [0, 0, 0, 0, 0, 0, 0, 0],
      route: "/functionalroles",
      verb: "GET"
    },
    {
      errorPercent: "0.000",
      errorsCount: 0,
      inThroughput: 0,
      latency50: 0,
      latency99: 0,
      outThroughput: 0,
      requests: 1,
      requestsPerSecond_dygraph: [
        [
          [new Date("2017-07-18T22:13:34.314Z"), 0],
          [new Date("2017-07-18T22:13:49.215Z"), 0],
          [new Date("2017-07-18T22:14:04.215Z"), 0],
          [new Date("2017-07-18T22:14:19.217Z"), 0],
          [new Date("2017-07-18T22:14:34.215Z"), 0],
          [new Date("2017-07-18T22:14:49.215Z"), 0],
          [new Date("2017-07-18T22:15:04.216Z"), 0],
          [new Date("2017-07-18T22:15:19.215Z"), 0],
          [new Date("2017-07-18T22:15:34.216Z"), 0]
        ],
        { labels: ["time", "Requests"] }
      ],
      requestsPerSecond_sparkline: [0, 0, 0, 0, 0, 0, 0, 0],
      route: "/ping",
      verb: "GET"
    }
  ],
  type: "Route"
};

let wrapper;

describe("RoutesGrid View", () => {
  beforeEach(() => {
    wrapper = mount(<RoutesGrid store={store} />).find("RoutesGrid");
  });

  test("matches snapshot", () => {
    const tree = renderer.create(<RoutesGrid store={store} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("renders correct components", () => {
    expect(wrapper.find("TableToolbar")).toHaveLength(1);
    expect(wrapper.find("Table")).toHaveLength(1);
    expect(wrapper.find("ErrorBoundary")).toHaveLength(1);
  });

  test("renders NotFoundError when there are no routes", () => {
    // create state with no metrics and reconfigure mock store with new state
    const state = Object.assign({}, mockState, { instance: { metrics: {} } });
    const store = configureMockStore()(state);
    wrapper = mount(<RoutesGrid store={store} />);
    expect(wrapper.find(NotFoundError)).toHaveLength(1);
  });

  test("passes the correct props down to TableToolbar", () => {
    expect(wrapper.find("TableToolbar").props()).toMatchObject({
      searchInputProps: {
        filterString: "",
        searchPlaceholder: "Search Routes"
      },
      sortByProps: {
        sortByAttribute: "route",
        sortByOptions: [
          { label: "Route", value: "route" },
          { label: "Requests", value: "requests" },
          { label: "Error %", value: "errorPercent" },
          { label: "Latency 50%", value: "latency50" },
          { label: "Latency 99%", value: "latency99" }
        ]
      }
    });
  });

  test("passes the correct props down to Table", () => {
    expect(wrapper.find("Table").props()).toMatchObject(routesTableProps);
  });
});
