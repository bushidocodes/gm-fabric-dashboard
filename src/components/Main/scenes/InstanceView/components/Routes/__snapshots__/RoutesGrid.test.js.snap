// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RoutesGrid View matches snapshot 1`] = `
Array [
  .c2 {
  background-image: url(magnifier.svg);
  background-repeat: no-repeat;
  background-size: contain;
  position: relative;
  text-indent: 20px;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  border: 1px solid rgba(0,0,0,0.25);
  border-radius: 3px;
  text-align: left;
  background-color: model:rgb;
  color: 255,255,255;
  valpha: 1;
  font-size: 14px;
  font-weight: 400;
  line-height: 1.1;
  width: 22em;
  -webkit-flex: 1 1 22em;
  -ms-flex: 1 1 22em;
  flex: 1 1 22em;
  padding: 6px 8px;
}

.c2::-webkit-search-cancel-button {
  cursor: pointer;
}

.c2::-webkit-input-placeholder {
  color: inherit;
  opacity: 1;
}

.c2::-moz-placeholder {
  color: inherit;
  opacity: 1;
}

.c2:-ms-input-placeholder {
  color: inherit;
  opacity: 1;
}

.c2::placeholder {
  color: inherit;
  opacity: 1;
}

.c5 {
  -webkit-flex-grow: 1;
  -ms-flex-grow: 1;
  flex-grow: 1;
  font-size: 14px;
  font-weight: 500;
  height: 28px;
  max-width: 125px;
  position: relative;
  width: 100%;
  z-index: 1020;
}

.c5 .Select-control {
  height: 28px !important;
}

.c5 .Select-control .Select-placeholder {
  line-height: 28px;
}

.c5 .Select-control .Select-value {
  line-height: 28px !important;
}

.c5 .Select-control .Select-value .Select-value-label {
  line-height: 28px;
}

.c5 .Select-control .Select-input {
  height: 28px;
}

.c5 + .c5 {
  margin-left: 8px;
}

.c6 {
  margin-left: 4px;
  opacity: 0.6;
  font-size: 14px;
}

.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: hsl(0,0%,96%);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  height: 36px;
  margin-left: auto;
  margin-right: auto;
  width: 100%;
  max-width: 1300px;
  padding: 2px 16px;
}

.c1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1 1 33%;
  -ms-flex: 1 1 33%;
  flex: 1 1 33%;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1 1 33%;
  -ms-flex: 1 1 33%;
  flex: 1 1 33%;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1 1 33%;
  -ms-flex: 1 1 33%;
  flex: 1 1 33%;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

@media all and (min-width:1300px) {
  .c0:after {
    background-color: inherit;
    content: "";
    height: inherit;
    left: 0;
    pointer-events: none;
    position: absolute;
    right: 0;
    top: 0;
    z-index: 1;
  }

  .c0 > * {
    position: relative;
    z-index: 2;
  }
}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <form>
        <input
          aria-label="Search Routes"
          class="c2"
          placeholder="Search Routes"
          type="search"
          value=""
        />
      </form>
    </div>
    <div
      class="c3"
    />
    <div
      class="c4"
    >
      <div
        class="Select c5 has-value is-searchable Select--single"
      >
        <input
          name="form-field-sort-by"
          type="hidden"
          value="route"
        />
        <div
          class="Select-control"
        >
          <span
            class="Select-multi-value-wrapper"
            id="react-select-3--value"
          >
            <div
              class="Select-value"
            >
              <span
                aria-selected="true"
                class="Select-value-label"
                id="react-select-3--value-item"
                role="option"
              >
                <span>
                  <span>
                    Sort 
                  </span>
                  <span
                    class="c6"
                  >
                    Route
                  </span>
                </span>
              </span>
            </div>
            <div
              class="Select-input"
              style="display:inline-block"
            >
              <input
                aria-activedescendant="react-select-3--value"
                aria-expanded="false"
                aria-haspopup="false"
                aria-owns=""
                role="combobox"
                style="box-sizing:content-box;width:5px"
                value=""
              />
              <div
                style="position:absolute;top:0;left:0;visibility:hidden;height:0;overflow:scroll;white-space:pre"
              />
            </div>
          </span>
          <span
            class="Select-arrow-zone"
          >
            <span
              class="Select-arrow"
            />
          </span>
        </div>
      </div>
    </div>
  </div>,
  .c15 {
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
}

.c14 {
  z-index: 1;
  height: 24px;
  width: 24px;
  color: currentColor;
  word-spacing: 0;
  vertical-align: middle;
}

.c14:not(:root) {
  overflow: visible;
}

.c10 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  min-height: 36px;
  -webkit-flex: 0 0 100%;
  -ms-flex: 0 0 100%;
  flex: 0 0 100%;
  cursor: pointer;
  -webkit-transition: all 0.2s ease;
  transition: all 0.2s ease;
  background-color: rgb(255,255,255);
  margin: 8px 0 0;
  border-radius: 3px;
  box-shadow: 0 0 0 1px hsl(0,0%,92%);
}

.c10:focus {
  outline: -webkit-focus-ring-color auto 5px;
}

.c10:focus:active:hover {
  outline: 0;
}

.c10 > * {
  padding-right: 16px;
  padding-top: 7px;
  padding-bottom: 7px;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.c10 > *:first-child {
  padding-left: 16px;
}

.c19 {
  -webkit-flex: 1 1 12%;
  -ms-flex: 1 1 12%;
  flex: 1 1 12%;
  text-align: left;
}

.c20 {
  color: rgb(10,171,42);
  -webkit-flex: 1 1 12%;
  -ms-flex: 1 1 12%;
  flex: 1 1 12%;
  text-align: left;
}

.c11 {
  -webkit-flex: 1 1 12%;
  -ms-flex: 1 1 12%;
  flex: 1 1 12%;
  text-align: left;
  -webkit-flex: 1 1 20%;
  -ms-flex: 1 1 20%;
  flex: 1 1 20%;
  position: relative;
}

.c18 {
  -webkit-flex: 1 1 12%;
  -ms-flex: 1 1 12%;
  flex: 1 1 12%;
  text-align: left;
  padding: 0;
  height: 24px;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c18 svg {
  -webkit-flex: 1 1 100%;
  -ms-flex: 1 1 100%;
  flex: 1 1 100%;
  border-radius: 3px;
  background-color: rgba(144,144,144,0.1);
}

.c16 {
  background-color: rgb(230,230,230);
  border-radius: 2px;
  bottom: 0;
  height: 1px;
  left: 16px;
  position: absolute;
  right: 16px;
}

.c17 {
  background-color: currentColor;
  border-radius: inherit;
  bottom: 0;
  color: rgb(10,171,42);
  left: 0;
  min-width: 2px;
  position: absolute;
  top: 0;
  width: 100%;
}

.c21 {
  background-color: currentColor;
  border-radius: inherit;
  bottom: 0;
  color: rgb(10,171,42);
  left: 0;
  min-width: 2px;
  position: absolute;
  top: 0;
  width: 0.005641748942172073%;
}

.c13 {
  font-size: 12px;
  font-weight: 500;
  text-align: center;
  line-height: 0.9;
  -webkit-letter-spacing: 0.06em;
  -moz-letter-spacing: 0.06em;
  -ms-letter-spacing: 0.06em;
  letter-spacing: 0.06em;
  height: fit-content;
  padding: 4px 4px;
  background-color: rgba(10,171,42,0.15000000000000002);
  border-radius: 3px;
  color: rgb(10,171,42);
  margin-right: 4px;
}

.c13 svg:first-child {
  margin: -100% -4px;
}

.c12 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 0 0 100%;
  -ms-flex: 0 0 100%;
  flex: 0 0 100%;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c12 > div {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  min-width: 0;
}

.c6 {
  background-color: rgba(0,0,0,0.8);
  border-radius: 6px;
  color: rgb(255,255,255);
  font-size: 14px;
  opacity: 0;
  padding: 10px;
  position: absolute;
  text-align: left;
  -webkit-transition: opacity 1s;
  transition: opacity 1s;
  visibility: hidden;
  white-space: normal;
  width: 160px;
  z-index: 1070;
  right: 120%;
  top: -5px;
}

.c5 {
  border-bottom: 1px solid rgb(217,217,217);
  cursor: help;
  display: inline-block;
  position: relative;
}

.c5:hover > :last-child {
  opacity: 1;
  visibility: visible;
  z-index: 1070;
}

.c0 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: rgb(255,255,255);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  font-weight: 500;
  overflow: visible;
  width: 100%;
}

.c9 {
  list-style: none;
  margin: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 0 8px 8px;
}

.c2 {
  -webkit-flex: 1 1 12%;
  -ms-flex: 1 1 12%;
  flex: 1 1 12%;
  text-align: left;
  padding-right: 16px;
  padding-top: 7px;
  padding-bottom: 7px;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.c2:first-child {
  padding-left: 16px;
}

.c3 {
  -webkit-flex: 1 1 12%;
  -ms-flex: 1 1 12%;
  flex: 1 1 12%;
  text-align: left;
  padding-right: 16px;
  padding-top: 7px;
  padding-bottom: 7px;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  overflow: visible;
}

.c3:first-child {
  padding-left: 16px;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c8 {
  font-size: 50%;
  color: rgb(77,77,77);
}

.c4 {
  padding: 8px;
}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
        style="flex:1 1 20%"
      >
        <span>
          Route
        </span>
      </div>
      <div
        class="c2"
      >
        <span>
          Requests/sec
        </span>
      </div>
      <div
        class="c2"
        style="text-align:right"
      >
        <span>
          Requests
        </span>
      </div>
      <div
        class="c2"
        style="text-align:right"
      >
        <span>
          Error %
        </span>
      </div>
      <div
        class="c3"
      >
        <div
          class="c4"
        >
          <div
            class="c5"
          >
            <span>
              Latency
            </span>
            <div
              class="c6"
            >
              Latency 50% refers to the average latency of the 50% percentile, while Latency 99% is the average latency for the slowest 1% of responses, also known as tail latency.
            </div>
          </div>
        </div>
        <div
          class="c7"
        >
          <div
            class="c8"
          >
            50%
          </div>
          <div
            class="c8"
          >
            99%
          </div>
        </div>
      </div>
    </div>
    <ol
      class="c9"
    >
      <li
        class="c10"
        role="link"
        tabindex="0"
      >
        <div
          class="c11"
        >
          <div
            class="c12"
          >
            <span
              class="c13"
            >
              <svg
                aria-labelledby="ariaLabelledby"
                class="c14"
                focusable="false"
                version="1.1"
                viewBox="0 0 24 24"
                xlink="http://www.w3.org/1999/xlink"
                xmlns="http://www.w3.org/2000/svg"
              >
                <g
                  class="c15"
                  fill="currentColor"
                  title=""
                >
                  <title />
                  <g
                    class="glyph"
                    fill="currentColor"
                    transform="translate(0 0) scale(1)"
                  >
                    <path
                      d="M11.5,13.940832 C9.01471863,13.940832 7,11.9261134 7,9.44083202 C7,6.95555065 9.01471863,4.94083202 11.5,4.94083202 C13.9852814,4.94083202 16,6.95555065 16,9.44083202 C16,11.9261134 13.9852814,13.940832 11.5,13.940832 Z M11.5521421,10.940832 C12.3805692,10.940832 13.0521421,10.2692591 13.0521421,9.44083202 C13.0521421,8.6124049 12.3805692,7.94083202 11.5521421,7.94083202 C10.723715,7.94083202 10.0521421,8.6124049 10.0521421,9.44083202 C10.0521421,10.2692591 10.723715,10.940832 11.5521421,10.940832 Z M8.38682576,13.9480614 C9.26316923,14.5385122 10.3188503,14.8831582 11.4550278,14.8831582 C12.5958345,14.8831582 13.6554876,14.535698 14.5339326,13.940832 L15.0304823,14.3285452 L11.4650086,17.0026504 L7.89953494,14.3285452 L8.38682576,13.9480614 Z"
                      id="get"
                    />
                  </g>
                </g>
              </svg>
              GET
            </span>
            /functionalroles
          </div>
          <div
            class="c16"
          >
            <div
              class="c17"
              width="100"
            />
          </div>
        </div>
        <div
          class="c18"
        >
          <svg
            preserveAspectRatio="none"
            viewBox="0 0 240 24"
          >
            <g>
              <circle
                cx="2"
                cy="12"
                r="2"
                style="stroke:currentColor;stroke-width:0;fill-opacity:.1;fill:none;pointer-events:auto"
              />
              <circle
                cx="35.714285714285715"
                cy="12"
                r="2"
                style="stroke:currentColor;stroke-width:0;fill-opacity:.1;fill:none;pointer-events:auto"
              />
              <circle
                cx="69.42857142857143"
                cy="12"
                r="2"
                style="stroke:currentColor;stroke-width:0;fill-opacity:.1;fill:none;pointer-events:auto"
              />
              <circle
                cx="103.14285714285714"
                cy="12"
                r="2"
                style="stroke:currentColor;stroke-width:0;fill-opacity:.1;fill:none;pointer-events:auto"
              />
              <circle
                cx="136.85714285714286"
                cy="12"
                r="2"
                style="stroke:currentColor;stroke-width:0;fill-opacity:.1;fill:none;pointer-events:auto"
              />
              <circle
                cx="170.57142857142858"
                cy="12"
                r="2"
                style="stroke:currentColor;stroke-width:0;fill-opacity:.1;fill:none;pointer-events:auto"
              />
              <circle
                cx="204.28571428571428"
                cy="12"
                r="2"
                style="stroke:currentColor;stroke-width:0;fill-opacity:.1;fill:none;pointer-events:auto"
              />
              <circle
                cx="238"
                cy="12"
                r="2"
                style="stroke:currentColor;stroke-width:0;fill-opacity:.1;fill:none;pointer-events:auto"
              />
              <polyline
                points="2 12 35.714285714285715 12 69.42857142857143 12 103.14285714285714 12 136.85714285714286 12 170.57142857142858 12 204.28571428571428 12 238 12 238 22 2 22 2 12"
                style="stroke:currentColor;stroke-width:0;fill-opacity:.1;fill:none;pointer-events:auto"
              />
              <polyline
                points="2 12 35.714285714285715 12 69.42857142857143 12 103.14285714285714 12 136.85714285714286 12 170.57142857142858 12 204.28571428571428 12 238 12"
                style="stroke:currentColor;stroke-width:1;stroke-linejoin:round;stroke-linecap:round;fill:none"
              />
            </g>
            <line
              style="stroke:grey;opacity:0.4"
              x1="2"
              x2="238"
              y1="14"
              y2="14"
            />
          </svg>
        </div>
        <div
          class="c19"
          style="text-align:right"
        >
          17,725
        </div>
        <div
          class="c20"
          style="text-align:right"
        >
          0%
        </div>
        <div
          class="c19"
          style="text-align:right;display:flex;flex-direction:column"
        >
          <div>
            0 ms
          </div>
          <div>
            9 ms
          </div>
        </div>
      </li>
      <li
        class="c10"
        role="link"
        tabindex="0"
      >
        <div
          class="c11"
        >
          <div
            class="c12"
          >
            <span
              class="c13"
            >
              <svg
                aria-labelledby="ariaLabelledby"
                class="c14"
                focusable="false"
                version="1.1"
                viewBox="0 0 24 24"
                xlink="http://www.w3.org/1999/xlink"
                xmlns="http://www.w3.org/2000/svg"
              >
                <g
                  class="c15"
                  fill="currentColor"
                  title=""
                >
                  <title />
                  <g
                    class="glyph"
                    fill="currentColor"
                    transform="translate(0 0) scale(1)"
                  >
                    <path
                      d="M11.5,13.940832 C9.01471863,13.940832 7,11.9261134 7,9.44083202 C7,6.95555065 9.01471863,4.94083202 11.5,4.94083202 C13.9852814,4.94083202 16,6.95555065 16,9.44083202 C16,11.9261134 13.9852814,13.940832 11.5,13.940832 Z M11.5521421,10.940832 C12.3805692,10.940832 13.0521421,10.2692591 13.0521421,9.44083202 C13.0521421,8.6124049 12.3805692,7.94083202 11.5521421,7.94083202 C10.723715,7.94083202 10.0521421,8.6124049 10.0521421,9.44083202 C10.0521421,10.2692591 10.723715,10.940832 11.5521421,10.940832 Z M8.38682576,13.9480614 C9.26316923,14.5385122 10.3188503,14.8831582 11.4550278,14.8831582 C12.5958345,14.8831582 13.6554876,14.535698 14.5339326,13.940832 L15.0304823,14.3285452 L11.4650086,17.0026504 L7.89953494,14.3285452 L8.38682576,13.9480614 Z"
                      id="get"
                    />
                  </g>
                </g>
              </svg>
              GET
            </span>
            /ping
          </div>
          <div
            class="c16"
          >
            <div
              class="c21"
              width="0.005641748942172073"
            />
          </div>
        </div>
        <div
          class="c18"
        >
          <svg
            preserveAspectRatio="none"
            viewBox="0 0 240 24"
          >
            <g>
              <circle
                cx="2"
                cy="12"
                r="2"
                style="stroke:currentColor;stroke-width:0;fill-opacity:.1;fill:none;pointer-events:auto"
              />
              <circle
                cx="35.714285714285715"
                cy="12"
                r="2"
                style="stroke:currentColor;stroke-width:0;fill-opacity:.1;fill:none;pointer-events:auto"
              />
              <circle
                cx="69.42857142857143"
                cy="12"
                r="2"
                style="stroke:currentColor;stroke-width:0;fill-opacity:.1;fill:none;pointer-events:auto"
              />
              <circle
                cx="103.14285714285714"
                cy="12"
                r="2"
                style="stroke:currentColor;stroke-width:0;fill-opacity:.1;fill:none;pointer-events:auto"
              />
              <circle
                cx="136.85714285714286"
                cy="12"
                r="2"
                style="stroke:currentColor;stroke-width:0;fill-opacity:.1;fill:none;pointer-events:auto"
              />
              <circle
                cx="170.57142857142858"
                cy="12"
                r="2"
                style="stroke:currentColor;stroke-width:0;fill-opacity:.1;fill:none;pointer-events:auto"
              />
              <circle
                cx="204.28571428571428"
                cy="12"
                r="2"
                style="stroke:currentColor;stroke-width:0;fill-opacity:.1;fill:none;pointer-events:auto"
              />
              <circle
                cx="238"
                cy="12"
                r="2"
                style="stroke:currentColor;stroke-width:0;fill-opacity:.1;fill:none;pointer-events:auto"
              />
              <polyline
                points="2 12 35.714285714285715 12 69.42857142857143 12 103.14285714285714 12 136.85714285714286 12 170.57142857142858 12 204.28571428571428 12 238 12 238 22 2 22 2 12"
                style="stroke:currentColor;stroke-width:0;fill-opacity:.1;fill:none;pointer-events:auto"
              />
              <polyline
                points="2 12 35.714285714285715 12 69.42857142857143 12 103.14285714285714 12 136.85714285714286 12 170.57142857142858 12 204.28571428571428 12 238 12"
                style="stroke:currentColor;stroke-width:1;stroke-linejoin:round;stroke-linecap:round;fill:none"
              />
            </g>
            <line
              style="stroke:grey;opacity:0.4"
              x1="2"
              x2="238"
              y1="14"
              y2="14"
            />
          </svg>
        </div>
        <div
          class="c19"
          style="text-align:right"
        >
          1
        </div>
        <div
          class="c20"
          style="text-align:right"
        >
          0%
        </div>
        <div
          class="c19"
          style="text-align:right;display:flex;flex-direction:column"
        >
          <div>
            0 ms
          </div>
          <div>
            0 ms
          </div>
        </div>
      </li>
    </ol>
  </div>,
]
`;
